File: apps/api/src/pinky_api/models/execution.py
import uuid
from datetime import datetime

from sqlalchemy import ForeignKey, Integer, String
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from pinky_api.models.base import Base, gen_uuid


class Execution(Base):
    __tablename__ = "executions"

    id: Mapped[uuid.UUID] = mapped_column(UUID, primary_key=True, default=gen_uuid)
    work_item_id: Mapped[uuid.UUID | None] = mapped_column(UUID, ForeignKey("work_items.id"))
    cluster_id: Mapped[uuid.UUID] = mapped_column(UUID, ForeignKey("cluster_registry.id"), nullable=False)
    execution_type: Mapped[str] = mapped_column(String, nullable=False)
    status: Mapped[str] = mapped_column(String, server_default="pending")
    started_at: Mapped[datetime | None] = mapped_column()
    completed_at: Mapped[datetime | None] = mapped_column()
    created_at: Mapped[datetime] = mapped_column(server_default="now()")


class ExecutionEvent(Base):
    __tablename__ = "execution_events"

    id: Mapped[uuid.UUID] = mapped_column(UUID, primary_key=True, default=gen_uuid)
    execution_id: Mapped[uuid.UUID] = mapped_column(UUID, nullable=False)
    event_type: Mapped[str] = mapped_column(String, nullable=False)
    sequence: Mapped[int] = mapped_column(Integer, nullable=False)
    payload: Mapped[dict] = mapped_column(JSONB, server_default="{}")
    occurred_at: Mapped[datetime] = mapped_column(nullable=False)


class Approval(Base):
    __tablename__ = "approvals"

    id: Mapped[uuid.UUID] = mapped_column(UUID, primary_key=True, default=gen_uuid)
    execution_id: Mapped[uuid.UUID] = mapped_column(UUID, ForeignKey("executions.id"), nullable=False)
    approver_id: Mapped[uuid.UUID | None] = mapped_column(UUID, ForeignKey("principals.id"))
    changeset_digest: Mapped[str] = mapped_column(String, nullable=False)
    target_resources: Mapped[list] = mapped_column(JSONB, nullable=False)
    status: Mapped[str] = mapped_column(String, server_default="pending")
    expires_at: Mapped[datetime] = mapped_column(nullable=False)
    created_at: Mapped[datetime] = mapped_column(server_default="now()")

File: packages/contracts/src/execution.ts
export type ExecutionStatus =
  | "pending"
  | "running"
  | "waiting_for_approval"
  | "completed"
  | "failed"
  | "timed_out"
  | "cancelled";

export type ExecutionEventType =
  | "started"
  | "progress"
  | "tool_used"
  | "approval_required"
  | "approval_granted"
  | "approval_rejected"
  | "completed"
  | "failed"
  | "verified"
  | "timed_out"
  | "rolled_back";

export interface Execution {
  id: string;
  work_item_id: string | null;
  cluster_id: string;
  cluster_display_name?: string;
  execution_type: string;
  status: ExecutionStatus;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
}

export interface ExecutionEvent {
  event_id: string;
  stream: string;
  aggregate_id: string;
  execution_id: string | null;
  type: ExecutionEventType;
  occurred_at: string;
  sequence: number;
  payload: Record<string, unknown>;
}

export interface TimelineEvent {
  id: string;
  execution_id: string;
  event_type: string;
  sequence: number;
  payload: Record<string, unknown>;
  occurred_at: string;
}

export interface RemediationStep {
  action: string;
  description: string;
  resource_kind: string;
  resource_namespace: string;
  resource_name: string;
  params: Record<string, unknown>;
  risk: "low" | "medium" | "high";
}

export interface Investigation {
  has_investigation: boolean;
  summary?: string;
  root_cause?: string;
  recommended_action?: string;
  confidence?: number;
  tool_calls?: string[];
  created_at?: string;
  remediation_steps?: RemediationStep[];
  manual_commands?: string[];
}

export type ApprovalStatus =
  | "pending"
  | "approved"
  | "rejected"
  | "expired"
  | "invalidated";

export interface Approval {
  id: string;
  execution_id: string;
  approver_id: string | null;
  changeset_digest: string;
  target_resources: Record<string, unknown>[];
  status: ApprovalStatus;
  expires_at: string;
  created_at: string;
}
